DataSources: refactor datasource pages to be reusable

leventebalogh opened the pull request:
### What changed?

**The main goal of these refactors were to be able to easily reuse pieces logic for displaying data-sources without duplicating the same setup logic.**

Refactored the **`app/features/datasources`** feature to:
- use Functional components instead of the Class components 
- reorganised the components under a `components/` and a `pages/` folder
- extracted out API related async functions to a `api.ts` module to make them easier to reuse and mock
- extracted out generic types to a `types.ts` module to make them easier to reuse
- started using `uid`s instead of `id`s for datasources wherever possible
- added utility hooks for interacting with information in the Redux store
- tried to simplify the logic in some places

leventebalogh committed:
refactor: move utility functions out of the redux actions

leventebalogh committed:
refactor: move password handlers to the feature root

leventebalogh committed:
refactor: move API related functions to an api.ts

leventebalogh committed:
refactor: move components under a /components folder

leventebalogh committed:
refactor: move page containers under a /pages folder and extract components

leventebalogh committed:
refactor: update mocks to be easier to reuse

leventebalogh committed:
refactor: move tests into a state/tests/ subfolder

leventebalogh committed:
refactor: expose 'initialState' for plugins

leventebalogh committed:
refactor: move generic types to the root folder of the feature

leventebalogh committed:
refactor: import path fixe

leventebalogh committed:
refactor: update import paths for app routes

leventebalogh committed:
chore: update betterer

leventebalogh committed:
refactor: fix type errors due to changed mock functions

leventebalogh committed:
chore: fix mocking context_srv in tests

leventebalogh committed:
refactor: udpate imports to be more concise

leventebalogh committed:
fix: update failing test because of mocks

leventebalogh committed:
refactor: use the new `navId` prop where we can

leventebalogh committed:
fix: use UID instead ID in datasource edit links

leventebalogh committed:
fix:clean up Redux state when unmounting the edit page

leventebalogh committed:
refactor: use `uid` instead of `id`

leventebalogh committed:
refactor: always fetch the plugin details when editing a datasource

The deleted lines could provide performance benefits, although they also make the
implementation more prone to errors. (Mostly because we are storing the information
about the currently loaded plugin in a single field, and it was not validating if it
is for the latest one).

We are planning to introduce some kind of caching, but first we would like to clean up
the underlying state a bit (plugins & datasources.

leventebalogh committed:
fix: add missing dispatch() wrapper for update datasource callback

ryantxu commented on public/app/features/datasources/components/tests/ButtonRow.test.tsx:
```diff
@@ -28,7 +20,7 @@ const setup = (propOverrides?: object) => {
 };
 
-describe('Button Row', () => {
+describe('<ButtonRow>', () => {
```
What is our best practice layout?  I have been adding tests to the *same* folder as the component, not a parallel test folder.

Should these tests all be in the components folder?

leventebalogh replied:
Sorry, answered it [here](https://github.com/grafana/grafana/pull/51874#issuecomment-1188661685).

leventebalogh replied:
Fixed it in https://github.com/grafana/grafana/pull/51874/commits/256d656b12bec45319425512f172ec8d171e1f07.

ryantxu commented on public/app/plugins/datasource/elasticsearch/configuration/mocks.ts:
```diff
@@ -1,19 +1,21 @@
 import { DataSourceSettings } from '@grafana/data';
 
-import { createDatasourceSettings } from '../../../../features/datasources/mocks';
+import { getMockDataSource } from '../../../../features/datasources/__mocks__';
```
nit
```suggestion
import { getMockDataSource } from 'app/features/datasources/__mocks__';
```

leventebalogh replied:
Thanks, nice catch! 👍

ryantxu committed:
Merge remote-tracking branch 'origin/main' into leventebalogh/data-connections-datasources

ryantxu commented:
@leventebalogh -- this looks great!

It is hard to tell what (if anything) actually changed given the style refactor.  However the changes you made need to be done eventually so 🤷‍♂️ 

I tested with a variety of datasources I use frequently and all seems unchanged, so that is good.

My only nit/question on the test structure is if we want special test folders, or just have them inline next to the component they test

ryantxu commented on public/app/features/datasources/components/ButtonRow.tsx:
```diff
@@ -14,7 +14,7 @@ export interface Props {
 }
 
-const ButtonRow: FC<Props> = ({ canSave, canDelete, onDelete, onSubmit, onTest, exploreUrl }) => {
+export function ButtonRow({ canSave, canDelete, onDelete, onSubmit, onTest, exploreUrl }: Props): React.ReactElement {
```
Genuine question... does the return type help with these components?  I though the implicit flavor ends up having more useful (and less verbose) types.

```suggestion
export function ButtonRow({ canSave, canDelete, onDelete, onSubmit, onTest, exploreUrl }: Props) {
```

leventebalogh replied:
Yeah, I cannot really think of any benefit of "being explicit" here either, I think TypeScript will just infer the return types fine. (The only scenario I could imagine is that it can prevent you from accidentally returning with something you don't want to).

The reason why I have added it to the new components was to be in sync with our [FE style guide](https://github.com/grafana/grafana/blob/main/contribute/style-guides/frontend.md#how-to-declare-functional-components), so we are a bit more coherent. What do you think?

leventebalogh replied:
Fixed in https://github.com/grafana/grafana/pull/51874/commits/f7dd31dc51804fd0a5df1a894ca07a01569ad42f. 👍

leventebalogh commented:
Thanks a lot for having a look at it @ryantxu and also for testing it out! 🙏 

Yeah, I know it's super hard to review these kind of changes, as mostly everything got shifted around, maybe I should have done it in batches (but then it would have taken much longer maybe).

**The location of the test files**
> My only nit/question on the test structure is if we want special test folders, or just have them inline next to the component they test.

Yeah, I don't have a strong opinion on this one, I think I just went with my taste, which doesn't necessarily mean that it is better (I liked grouping them in a subfolder, as like this they are still "next to" the component they are testing, but it is also easier to scan the actual component files to find something). I am happy to move them next to the modules they are testing though.

 **Absolute vs relative paths**
> Also rather than ../../../features I like the absolute path app/features/...

I agree, I prefer the absolute ones, too. My thinking was to use absolute paths everywhere "outside" of a feature (e.g. app/features/datasources) and use relative paths when referencing something from inside the same feature. What do you think?

leventebalogh committed:
refactor: prefer using absolute import paths

Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

leventebalogh committed:
fix: ESLINT import order issue

ryantxu commented:
>  I am happy to move them next to the modules they are testing though.

Lets do that -- discussed with torkel, and that is the pattern that seems better.

Other than the style nits, I think this is ready to merge

leventebalogh committed:
refactor: put test files next to their files

leventebalogh committed:
refactor: use implicit return types for components

leventebalogh committed:
Merge remote-tracking branch 'origin/main' into leventebalogh/data-connections-datasources

* origin/main: (36 commits)
  Dashboard: Fix iteration property change triggering unsaved changes warning  (#51272)
  Public Dashboards: Add Public Tag to Dashboard Title (#52351)
  Secrets: Add logging to track secrets migration (#52481)
  Adds documentation for public dashboards under the dashboards section. (#52126)
  Update upgrade-grafana.md (#52406)
  FolderPage: Improve folder page to work with new nav breadcrumbs and modify Dashboard page show path based breadcrumbs"  (#52428)
  Alerting: Rename sender.Sender to sender.ExternalAlertmanagers (#52463)
  CloudWatch: fix log explorer context (#52118)
  Secret migration from Sql KV Store to Secret Plugin (#52191)
  Tree navigation in inline editor (#52427)
  StyleGuide: Update frontend style guide (#52471)
  Alerting: Fix Slack push notifications (#52391)
  Loki: Fix adding of multiple label filters when parser (#52335)
  Alerting: Preserve new-lines from custom email templates in rendered email (#52253)
  Scene: Fixing state issue with useState when SceneObject instance changes (#52372)
  Scenes: Improve typing of scene state to avoid type guards and casting (#52422)
  ServiceAccounts: Updates the service accounts list page to look good in new top nav design (#52425)
  Dashboard Links: Fix styles for very long dashboard titles (#52443)
  Storage: remove orgId from sql config (#52426)
  Docker: Install `git` in `grafana/grafana-ci-deploy` image (#52466)
  ...

leventebalogh committed:
fix: remove caching from datasource fetching

I have introduced a cache to only fetch data-sources once, however as we
are missing a good logic for updating the instances in the Redux store
when they change (create, update, delete), this approach is not keeping the UI in sync.
Due to this reason I have removed the caching for now, and will reintroduce it once we have a
more robust client-side state logic.
